### ISSUE TITLE
4.13.0 release checklist

### ISSUE THREAD

**Issue by corinagum** [microsoft/BotFramework-WebChat#3826]
# Checklist

## Build

1. [x] ~Bump MockBot to Bot Framework SDK release 4.13.0~
   - https://www.npmjs.com/package/botbuilder/v/4.13.0
   - (We will not bump for patch release)
1. [x] ~Bump `botframework-directlinejs` to `0.14.1`~
   - No newer version
1. [x] Bump to `4.13.0`
   - [x] Update `CHANGELOG.md` to mark specific changes in `4.13.0`
   - [x] Run `npm version --no-git-tag-version 4.13.0`
   - [x] Merged into `master`, the PR number is #3827
   - Commit is `585da0a`
   - Do not merge any other unrelated changes after this PR. Any other PR merged, will need to be re-tested
1. [x] Run daily pipeline manually, set "generate release version number" to `true`
   - (This will not push to NPM or CDN)
   - Pipeline name is `BotFramework-WebChat-daily`
   - The build number is [`230592`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=230592&view=results) and commit is `585da0a`
1. [x] Wait for [`WebChat-release-testing` pipeline](https://fuselabs.visualstudio.com/BotFramework-WebChat/_release?_a=releases&view=mine&definitionId=17) to complete
   - Pipeline name is `Push-Release-Testing-to-GitHub-Pages`
   - The release ID is [`2697`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2697)
1. [x] Add manual tests to WebChat-release-testing as needed

## Test

> The test should run against the build artifacts from Azure Pipelines.

1. [x] Manual testing on major browsers using `webchat-release-testing`
   - [x] Before starting testing, update all the browser version to latest
   - [x] Chrome 89.0.4389.114
   - [x] Edge 89.0.774.68
   - [x] Firefox 87.0
   - [x] IE11 (Windows 10 21343.1010)
   - [x] macOS Safari 14.0.3 (16610.4.3.1.7)
   - [x] iOS Safari 14.4.2
   - [x] Android Chrome 89.0.4389.105
1. [x] Test specific fixes related to `4.13.0` and previous releases
   
   - Markdown external link indicator (visual and AT)

## Release
1. [x] Verify on WebChat-release-testing
1. [x] Make sure you are on `master` ~or `qfe`~ branch, run `git status` to check
1. [x] `git pull`
1. [x] Verify `/package.json`, `/package-lock.json`, and `CHANGELOG.md` has a version of `4.13.0`
1. [x] `git log`
   - Verify the latest commit is `585da0a`
1. [x] `git tag v4.13.0`
1. [x] `git push -u upstream v4.13.0`
   - You do not need to kick off a build again, use the previous build
1. [x] Create a new GitHub release, copy entries from `CHANGELOG.md`
   - [x] Subresource Integrity can be generated by
      - From CDN: `curl -H 'Accept-Encoding: gzip' https://cdn.botframework.com/botframework-webchat/4.13.0/webchat.js | gunzip - | openssl dgst -sha384 -binary | openssl base64 -A`
      - From local: `cat webchat.js | openssl dgst -sha384 -binary | openssl base64 -A`
   - [x] Attach assets including 3 JS files, `stats.json` and 5 tarballs
      - You can copy the artifacts from `webchat-release-testing/drops`
      - Tarballs download from npmjs
         ```
         curl -L -o botframework-directlinespeech-sdk-4.13.0.tgz https://registry.npmjs.org/botframework-directlinespeech-sdk/-/botframework-directlinespeech-sdk-4.13.0.tgz
         curl -L -o botframework-webchat-4.13.0.tgz https://registry.npmjs.org/botframework-webchat/-/botframework-webchat-4.13.0.tgz
         curl -L -o botframework-webchat-core-4.13.0.tgz https://registry.npmjs.org/botframework-webchat-core/-/botframework-webchat-core-4.13.0.tgz
         curl -L -o botframework-webchat-api-4.13.0.tgz https://registry.npmjs.org/botframework-webchat-api/-/botframework-webchat-api-4.13.0.tgz
         curl -L -o botframework-webchat-component-4.13.0.tgz https://registry.npmjs.org/botframework-webchat-component/-/botframework-webchat-component-4.13.0.tgz
         ```
1. [x] Kick off release to NPM
   - Release name is `[[PROD]]Push-WebChat-to-npmjs`
   - The build number is [230592](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=230592&view=results) release number is `32` and commit is `585da0a`
   - Retain the build indefinitely
1. [x] Kick off release to CDN (CDN CUTOFF TIME IS 2PM PST)
   1. [x] Prepare the email for approval
   - Release name is `[[PROD]]Push-WebChat-to-Prod-CDN-with-approval`
   - The build number is [`230592`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=230592view=results), release number is [`11`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=xxxx) and commit is `abc123`
   - Script build number is `131156` (this is fixed)
   1. [x] Send reminder email to approvers
   - Retain the build indefinitely

## Post-release verification - complete within 30mins / 1hr of the release

- [x] Test using `webchat-release-testing`
   1. [x] Clone https://github.com/corinagum/WebChat-release-testing/
   1. [x] `01.create-react-app`
      1. [x] Nuke `01.create-react-app/node_modules`
      1. [x] `npm install`
      1. [x] `npm install botframework-webchat@4.13.0` (just install the bundle package)
      1. [x] `npm run build`
   1. [x] Others
      -  Using script tags from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.13.0, with subresource integrity
         ```
         <script
            crossorigin="anonymous"
            integrity="sha384-"
            src="https://cdn.botframework.com/botframework-webchat/latest/webchat-es5.js"
         ></script>
         ```
   1. [x] `npx serve` (at repo root)
   1. [x] Go to http://localhost:5000/ to test, including IE11

## Notification to interested parties

- [x] Update partner page on Adaptive Cards doc
   - https://docs.microsoft.com/en-us/adaptive-cards/resources/partners
   - PR at https://github.com/MicrosoftDocs/AdaptiveCards/pull/381
- [ ] Notify related parties for the following fixes
   - [ ] SDK team
   - [ ] Omnichannel
   - [ ] Pooja
- [ ] Update root README.md with feature notes -- Note: PR will be combined with post-release checklist PR
- [x] Update https://github.com/microsoft/BotFramework-sdk/issues/6245
   
---

# Post-release checklist

These are chores that we should do before starting the cycle to reduce ripple effects if we do it in mid-cycle.

Tips: 

- Clean your repo before start
- Remove `node_modules` from all folder
   - `git clean -fdx`
- Never delete `package-lock.json`
- If you mess it up, tableflip and redo
- In `component/package.json`
   - Remove reference to `botframework-webchat-core` by hand-modifying `package.json`
   - Then, `npm install` (symlinks will be broken afterward)
   - Then, add those references back by hand-modifying `package.json`
   - This also applies for other packages with similar dependencies/symlinks
   - To build afterward, do tableflip to rebuild those symlinks

## Applies to all releases

> This list should be copied to versions in the future.

- [x] ~If on QFE branch, make sure `CHANGELOG.md` and version number bump is cherry-picked to `master`~
   - [ ] ~`git checkout master`~
   - [ ] ~`git cherry-pick XXX` (the commitish for bumping version number and `CHANGELOG.md`)~
- [x] Correct the date for 4.13.0 in `CHANGELOG.md`
   - There could be last minute fixes that could push the planned date later than the one in `CHANGELOG.md`
- [x] Bump `package.json` to `4.13.1-0`
   - Run `npm version prepatch --no-git-tag-version`
- [x] Update [`servicingPlan.json`](https://github.com/microsoft/BotFramework-WebChat/blob/master/packages/embed/servicingPlan.json)
   - Add deprecation notes for previous versions
   - Subresource integrity hash from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.13.0
- [x] Update all samples to use `4.13.0`
   - Some samples are pointing to GitHub Releases because the sample need new features from daily build
   - Search "https://github.com/microsoft/BotFramework-WebChat/releases/download/"
      - And replace with "https://cdn.botframework.com/botframework-webchat/latest/"

The PR is https://github.com/microsoft/BotFramework-WebChat/pull/XXX

### Bump all dependencies to latest version

On each `package.json` under root and `/packages/`:

- [x] All NPM packages (`@babel/*`, `jest`, `lerna`, `typescript`, `webpack`, etc) in PR #3831
   - Don't bump the following packages
      - `react`, `react-dom`, `redux`, `redux-saga`, `react-film`, `react-scroll-to-bottom`, `react-say`, `react-dictate-button`, `@types/react`
      - `remark` and `strip-markdown`, `source-map-loader` (because of webpack)
         - Newer `remark` don't work under IE11
         - Please go to the their repos and see if this is fixed or by design
      - `sanitize-html@1`
         - Starting from version 2, they included `nanoid` deps which does not work in ES5 environment
         - compulim: Seems `sanitize-html` removed `nanoid` from their dependencies, please try to bump it. We use `sanitize-html` while rendering Markdown
      - `serve`
      - `rxjs` on `playground`
      - `webpack@4`
         - Starting from version 5, they dropped Node.js polyfills for browsers
         - Please file an issue: Bump `webpack@5`
         - Keep the following dependencies because their latest version requires `webpack@5`
            - [`copy-webpack-plugin@6`](https://npmjs.com/package/copy-webpack-plugin)
            - [`html-webpack-plugin@4`](https://npmjs.com/package/html-webpack-plugin)
      - `microsoft-cognitiveservices-speech-sdk`
      - `adaptivecards`
        - We will bump it in independent PR
      - `expect@25`
        - The newer version requires Node.js (`graceful-fs`)
- [x] Run `npm audit fix` to make sure everything is fixed
- [x] Test under IE11 to make sure all dependencies are working

### Bump Docker image

> The Docker image can be found at root `docker-compose.yml` and `Dockerfile*`.

- [ ] Docker container for headless Chrome (#XXX)
   - They recently moved from `3.14.159-xxx` tag scheme to a more sensible `87.0` tag scheme
   - Tags can be found at https://hub.docker.com/r/selenium/node-chrome/tags
   - Preferably in separate PR because screenshots change can be large occasionally
   - Run tests locally, as the screenshots can be slightly different
- Note on 11/10/2020: `4.0.0-alpha-7-prerelease-20200907` is still prerelease; therefore not bumping (corinagum)

**Comment by corinagum** [comment]
https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.13.0
